### PR TITLE
Add OrderdProduct module

### LINKED ISSUE #59: Add OrderedProduct module
- Create OrderedProduct  module - fields and features.
- Write a test data.
- Write a pytest tests. 

Done on PR #58

### PR DESCRIPTION
**Pull request by MeitarRizner**
Add OrderdProduct module with all the features and tests. 

Fixes #59 

All the classes in models file are temporary except OrderedProduct (These classes didn't merge yet).

*Please let me know about missing features and missing tests*

### CODE REVIEW

**Comment by pazMenachem** [comment]
since im not exactly sure which basic tests the mentors/maintainers wants for module, the code looks good to me.
also assuming all the base class (user supplier client etc) will be deleted once those classes will be merged.

**Comment by Omeramsc** [comment]
Why does this PR also creates other modules? is it a WIP? or depends on another PR?

File: ordered_product/migrations/0001_initial.py
Comment (thetaxas2468) [inline ordered_product/migrations/0001_initial.py:2]: Delete comment

File: ordered_product/models.py
Comment (thetaxas2468) [inline ordered_product/models.py:65]: Django creates auto field primary key,delete this one

File: ordered_product/models.py
Comment (thetaxas2468) [inline ordered_product/models.py:101]: You have increase,decrease,order(with quantity = 1) functions that needs to needs to be here and missing,Look at the DFD

**Comment by thetaxas2468** [comment]
link this PR to an issue
check out :
https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

File: ordered_product/models.py
Comment (yftacherzog) [inline ordered_product/models.py:64]:
```suggestion
class OrderedProduct(models.Model):
```

File: ordered_product/models.py
Comment (yftacherzog) [inline ordered_product/models.py:72]: change variable and function name to snake_case with lowercase only. Also for the rest of the PR.

File: ordered_product/tests.py
Comment (yftacherzog) [inline ordered_product/tests.py:50]: You're creating fixtures for (supplier, client, etc.) but you're not using them. I was expecting to see them being used in this fixture instead of creating yet more instances. You can also save each of them from within those fixtures rather than cluttering this one.

File: ordered_product/tests.py
Comment (yftacherzog) [inline ordered_product/tests.py:94]: You can just use the same OrderedProduct again and keep the test cleaner.

File: ordered_product/tests.py
Comment (yftacherzog) [inline ordered_product/tests.py:113]: Why do we need to `list()` the result at the positive test but not on the negative test?

File: ordered_product/tests.py
Comment (yftacherzog) [inline ordered_product/tests.py:115]: Why using `in` here and `==` on the previous one? I think `==` is more appropriate (had a similar discussion with @thetaxas2468)

**Comment by thetaxas2468** [comment]
> link this PR to an issue check out : https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

Still not linked checkout the link, it should be something like this:
Commit information with an opened issue
<9 options> the issue id

Example:
Information
fixes #59

**Comment by MeitarRizner** [comment]
> Why does this PR also creates other modules? is it a WIP? or depends on another PR?

Its depends on another PR.

File: ordered_product/tests.py
Comment (MeitarRizner) [reply ordered_product/tests.py:50]: I tried this (use fixture as instance) and it didn't work. I will try again

Comment (yftacherzog) [reply ordered_product/tests.py:50]: Let us know what's not working and we can try and figure this out together.

File: ordered_product/admin.py
Comment (pazMenachem) [inline ordered_product/admin.py:9]:
```suggestion
admin.site.register(OrderedProduct, OrderedProductAdmin)
```

File: ordered_product/admin.py
Comment (pazMenachem) [inline ordered_product/admin.py:2]:
```suggestion
from ordered_product.models import OrderedProduct
```

File: ordered_product/migrations/0002_test_data.py
Comment (pazMenachem) [inline ordered_product/migrations/0002_test_data.py:2]:
```suggestion
from ordered_product.models import Supplier, Client, Product, SupplierProduct, DeliveryLocation, OrderedProduct
```

**Review by pazMenachem — COMMENTED** [review]
take note that User abstract class wont have user type field any more, therefor change the relevant sections in your code that include Supplier/Client.

File: ordered_product/models.py
Comment (Omeramsc) [inline ordered_product/models.py:8]: Are all the classes in this file are temporary? because we don't need to redefine them.
if so, please mention it in the description.

File: ordered_product/models.py
Comment (Omeramsc) [inline ordered_product/models.py:89]: no need for this wrapper function, it doesn't add anything.

File: ordered_product/tests.py
Comment (thetaxas2468) [inline ordered_product/tests.py:127]: test your quantity validator

File: ordered_product/models.py
Comment (yftacherzog) [inline ordered_product/models.py:66]: snake_case

File: ordered_product/models.py
Comment (yftacherzog) [inline ordered_product/models.py:67]: snake_case

File: ordered_product/models.py
Comment (yftacherzog) [inline ordered_product/models.py:70]: snake_case

File: ordered_product/models.py
Comment (yftacherzog) [inline ordered_product/models.py:74]: snake_case

File: ordered_product/models.py
Comment (yftacherzog) [inline ordered_product/models.py:78]: snake_case

File: ordered_product/models.py
Comment (yftacherzog) [inline ordered_product/models.py:66]: Can't we use `get` here? Do we need filter functionality for this field?

File: ordered_product/tests.py
Comment (yftacherzog) [inline ordered_product/tests.py:60]: It would make sense to assert `not in` before adding.

File: ordered_product/tests.py
Comment (yftacherzog) [inline ordered_product/tests.py:70]: All your tests start with those 2 lines. I would make sense to put that inside a fixture.

File: ordered_product/tests.py
Comment (yftacherzog) [inline ordered_product/tests.py:74]: For all those tests, `==` will make better sense comparing to `in`. This way, you will assert that you have exactly what you were filtering for and not other stuff.

File: ordered_product/tests.py
Comment (yftacherzog) [inline ordered_product/tests.py:108]: Why are we saving this one?

File: ordered_product/tests.py
Comment (yftacherzog) [inline ordered_product/tests.py:113]: Is this where we plan to fail? What's actually failing here?

File: ordered_product/tests.py
Comment (yftacherzog) [inline ordered_product/tests.py:124]: A lot of magic numbers here, which makes the test more fragile and less transparent. I'd suggest either having all values set here (e.g. assign 3, decrease by 1, assert 2, and so on), or make the values proportional to existing value (e.g. store the current value to a local var, decrease by 1, assert equals temp-1, decrease by temp+8, assert deleted, etc.).

File: ordered_product/tests.py
Comment (yftacherzog) [inline ordered_product/tests.py:133]: same comment about magic numbers.

File: ordered_product/tests.py
Comment (thetaxas2468) [inline ordered_product/tests.py:50]: In here your tests pass because in your 0002 test file you have the same product,supplier,supplierproduct,deliverylocation (if you change the fields value i think all your tests will fail)

so before returning the ordered product save each one of the parameters deliver_location0....... by calling the functions that we have created in our models and after that return the OrderedProduct

File: ordered_product/tests.py
Comment (thetaxas2468) [inline ordered_product/tests.py:72]: What are you testing here?

File: ordered_product/tests.py
Comment (thetaxas2468) [inline ordered_product/tests.py:81]: What are you testing here?

File: ordered_product/tests.py
Comment (thetaxas2468) [inline ordered_product/tests.py:90]: this should be in its own test function i guess

File: ordered_product/tests.py
Comment (thetaxas2468) [inline ordered_product/tests.py:99]: this should be in its own test function i guess

**Comment by thetaxas2468** [comment]
Please rebase your commit on top of upstream/main and fix your migrations because supplier is already merged and deliverylocation is about to be merged also and fix your paths in order look at my review on other PRs,Rawan and Lior

File: ordered_product/models.py
Comment (MeitarRizner) [reply ordered_product/models.py:66]: You are right. I'll fix it.

File: ordered_product/models.py
Comment (pazMenachem) [inline ordered_product/models.py:103]: maybe check whether qn is 0, since we agreed product with 0 quantity is not valid.
would suggest raising an error if qn is 0 or create other indication the instance hasnt been saved.

**Review by pazMenachem — COMMENTED** [review]
there are methods that do the same as built in methods, no point warping those methods as  i see it.

File: ordered_product/tests.py
Comment (MeitarRizner) [reply ordered_product/tests.py:113]: I'll fix it.

File: ordered_product/models.py
Comment (thetaxas2468) [reply ordered_product/models.py:103]: if qn <= 0 and not only 0

File: ordered_product/tests.py
Comment (MeitarRizner) [reply ordered_product/tests.py:50]: I'm using the supplier0 that Paz wrote and the tests still pass.. So I don't need to do these changes.

File: ordered_product/tests.py
Comment (MeitarRizner) [reply ordered_product/tests.py:72]: I deleted this filter

File: ordered_product/tests.py
Comment (MeitarRizner) [reply ordered_product/tests.py:81]: I created a new fixture 'client1' to filter by this fixture instead of ("")

File: ordered_product/tests.py
Comment (MeitarRizner) [reply ordered_product/tests.py:90]: Why? I'm don't understand

File: ordered_product/tests.py
Comment (MeitarRizner) [reply ordered_product/tests.py:99]: Why? I'm don't understand

File: ordered_product/models.py
Comment (MeitarRizner) [reply ordered_product/models.py:103]: Okay

**Comment by MeitarRizner** [comment]
> there are methods that do the same as built in methods, no point warping those methods as i see it.

Which methods?

File: ordered_product/tests.py
Comment (yftacherzog) [inline ordered_product/tests.py:55]: You should not have try/except blocks inside your tests. You should instead use the `with pytest.raises(ValidationError)` pattern. It's already implemented in several other PRs, so have a look.

File: ordered_product/tests.py
Comment (yftacherzog) [inline ordered_product/tests.py:84]: Use `with pytest.raises(...)` instead of `try/except`.

File: ordered_product/tests.py
Comment (yftacherzog) [inline ordered_product/tests.py:81]: The negative scenario should be in a test case of its own.

File: ordered_product/tests.py
Comment (yftacherzog) [inline ordered_product/tests.py:54]: make the context manager include only the line that you expect to generate the exception. This way you increase the chances for actually spotting the error you were expecting.
```suggestion
        temp_ordered_product = OrderedProduct(
                delivery_location_id=delivery_location0,
                user_name=client0,
                supplier_product_id=supplier_product0,
                quantity=-2)
        with pytest.raises(ValidationError):
            temp_ordered_product.save_ordered_product()
        assert temp_ordered_product not in OrderedProduct.objects.all()
```

File: ordered_product/tests.py
Comment (Omeramsc) [inline ordered_product/tests.py:64]: what do we test here?

Comment (MeitarRizner) [reply ordered_product/tests.py:64]: > what do we test here?

When clients decrease quantity by more than they ordered so the order needs to deleted (if the quantity less or equals to zero --> delete order)

File: ordered_product/tests.py
Comment (Omeramsc) [inline ordered_product/tests.py:82]: This assertion seems redundant. you created those products above, you should be aware of what they consist of.

Comment (Omeramsc) [reply ordered_product/tests.py:82]: (same for all of the assertions below)

File: ordered_product/tests.py
Comment (Omeramsc) [inline ordered_product/tests.py:77]: what is the difference between client1 and client zero? why do we need them both?

**Comment by Omeramsc** [comment]
Do note that when the other 2 open PRs will merge, you'll have to rebase and remove the other classes' declarations

File: ordered_product/tests.py
Comment (MeitarRizner) [reply ordered_product/tests.py:82]: Okay. I'll fix it.

File: ordered_product/tests.py
Comment (MeitarRizner) [reply ordered_product/tests.py:77]: I'm understanding it now.. I'll change the test.

File: ordered_product/models.py
Comment (yftacherzog) [inline ordered_product/models.py:50]: the `save` line is potentially putting the DB in an invalid state until you delete it. it means a race condition in which others can access invalid data between the time you save it and the time you delete it.

Honestly, any way you turn this around, there's going to be some sort of race condition here unless you make this method atomic (e.g. you decrement a value different than the one you read, etc.), and I don't want you to spend time on that. so, for now, I think you can leave it as-is, but if someone asks you about it one day, just know there's an issue here :)

Comment (yftacherzog) [reply ordered_product/models.py:50]: (All in all, nothing to do here for now, just FYI for whoever reads)

**Review by yftacherzog — APPROVED** [review]
lgtm

**Review by thetaxas2468 — APPROVED** [review]
LGTM!

**Review by Omeramsc — APPROVED** [review]
LGTM, but we'll have to wait for #38 to merge, and rebase this on it, before we'll be able to merge.

**Comment by Omeramsc** [comment]
@MeitarRizner #38 is merged, solve the conflicts and we'll be able to merge it :)

**Comment by MeitarRizner** [comment]
> @MeitarRizner #38 is merged, solve the conflicts and we'll be able to merge it :)

@Omeramsc Done.

**Review by Omeramsc — APPROVED** [review]
Good job! merged

**Pull request merged**
